packages/react-components/src/box/src/Box.tsx

```diff
@@ -0,0 +1,46 @@
+import { ComponentProps, ElementType, ForwardedRef, ReactElement, ReactNode } from "react";
+import { forwardRef, omitProps } from "../../shared";
+
+export interface InnerBoxProps {
+    /**
+    * An HTML element type or a custom React element type to render as.
+    */
+    as?: ElementType;
+    /**
+     * [Slot](?path=/docs/getting-started-slots--page) to render into.
+     */
+    slot?: string;
+    /**
+     * React children
+    */
+    children?: ReactNode;
+    /**
+    * @ignore
+    */
+    forwardedRef: ForwardedRef<any>;
+}
+
+export function InnerBox(props: InnerBoxProps): ReactElement {
+    const {
+        as: Wrapper = "div",
+        children,
+        forwardedRef,
+        ...rest
+    } = omitProps(props, ["slot"]);
+
+    return (
+        <Wrapper
+            {...rest}
+            ref={forwardedRef}
+        >
+            {children}
+        </Wrapper>
+    );
+}
+export const Box = forwardRef<InnerBoxProps>((props, ref) => (
+    <InnerBox {...props} forwardedRef={ref} />
+));
```

Comment:
Using the forwardRef from Orbit, we get code that is really similar to regular react. 

Here are some exemple with the types generated

 **Generic use case : We don't specify a type**

```
export const Box = forwardRef<InnerBoxProps>((props, ref) => (
    <InnerBox {...props} forwardedRef={ref} />
));


export type BoxProps = ComponentProps<typeof Box>;

type OnClick = BoxProps["onClick"]; // (event: MouseEvent<Element, MouseEvent>) => void
type UnknownProperty = BoxProps["aaaa"]; // any, we allow extra properties, but we don't type them
type Slot = BoxProps["slot"]; //  string
type Ref = BoxProps["ref"]; //  ((instance: InnerBoxProps<Element>) => void) | RefObject<InnerBoxProps<Element>>

```
So we see that when we don't specifiy a type as the 2nd generic parameter in forward Ref, we extends every html properties, using with an Element type.


**Specific properties**
```
export const BoxAsInput = forwardRef<InnerBoxProps, "input">((props, ref) => {
    return <InnerBox {...props} forwardedRef={ref} />;
});

export type BoxProps = ComponentProps<typeof BoxInput>;

type OnClick = BoxProps["onClick"]; // (event: MouseEvent<HTMLInputElement, MouseEvent>) => void
type UnknownProperty = BoxProps["aaaa"]; // any, we allow extra properties, but we don't type them
type Slot = BoxProps["slot"]; //  string
type Ref = BoxProps["ref"]; //  ((instance: HTMLInputElement) => void) | RefObject<HTMLInputElement>
type OnChange = BoxProps["onChange"]; // (event: ChangeEvent<HTMLInputElement>) => void

```

**Custom Components**

```

const MyRouterLink = React.forwardRef<HTMLLinkElement, { allo: boolean }>((props, ref) => {
    return <>{props.allo}</>;
});

export const Box = forwardRef<InnerBoxProps, typeof MyRouterLink>((props, ref) => (
    <InnerBox {...props} forwardedRef={ref} />
));
export type BoxProps = ComponentProps<typeof Box >;

type OnClick = BoxProps["onClick"]; //any -> since we provided a component, we don't add all html props
type UnknownProperty = BoxProps["aaaa"]; // any, we allow extra properties, but we don't type them
type Slot = BoxProps["slot"]; //  string
type Ref = BoxProps["ref"]; // ((instance: HTMLLinkElement) => void) | RefObject<HTMLLinkElement>
type CustomProps = BoxProps["allo"]; //  boolean
```

Reply:
So the following would also work with full typing available? (giving we do exposed BoxProps as you suggested)

```
export const ParagraphBox = forwardRef<BoxProps, "p">((props, ref) => (
    <Box {...props} ref={ref} />
));
```

The best way to do this would be the following but both syntax would be supported right?

```
<Box as="p">Blabla</Box>
```